src/components/lib/posix/posix.c

```diff
@@ -331,124 +346,33 @@ struct futex_data futexes[FUTEX_COUNT];
 struct futex_data *
 lookup_futex(int *uaddr)
 {
-	int last_free = -1;
-	int i;
-
-	for (i = 0; i < FUTEX_COUNT; i++) {
-		if (futexes[i].uaddr == uaddr) {
-			return &futexes[i];
-		} else if (futexes[i].uaddr == 0) {
-			last_free = i;
-		}
-	}
-	if (last_free >= 0) {
-		futexes[last_free] = (struct futex_data) {
-			.uaddr = uaddr
-		};
-		ps_list_head_init(&futexes[last_free].waiters);
-		return &futexes[last_free];
-	}
-	printc("Out of futex ids!");
 	assert(0);
 }
 
 /* TODO: Cleanup empty futexes */
 
-struct sl_lock futex_lock = SL_LOCK_STATIC_INIT();
+//struct sl_lock futex_lock = SL_LOCK_STATIC_INIT();
 
 /*
  * precondition: futex_lock is taken
  */
 int
 cos_futex_wait(struct futex_data *futex, int *uaddr, int val, const struct timespec *timeout)
 {
```

Comment:
This is all pretty challenging logic that Gregor spent a lot of time on. What I'd like: create an issue with a link to the commit in which this is removed, so that we can know how to easily bring back this code when hook up posix to the rest of the system.

Reply:
I think there have to be ways to do this that don't sacrifice correctness. (ie that fault instead of being unsafe concurrently) I'm always nervous leaving things in a subtly broken state--it's easy to forget the exact brokenness and put yourself in debugging hell for no reason later

Reply:
Great point.

Perhaps we replace the current lock implementation with something that simply errors out on contention:

```c
struct lock {
        unsigned long owner;
};
struct lock global_lock = (struct lock) { .owner = 0 };
void
lock_take(void) {
        // failure here indicates that you're using concurrency/parallelism, which posix doesn't yet support!
        if (!ps_cas(&global_lock.owner, 0, cos_thdid())) assert(0);
}
void
lock_release(void) {
        // failure here indicates unpaired take and release.
        assert(global_lock.owner == cos_thdid());
        ps_store(&global_lock.owner, 0);
}
```
Replace normal lock behavior which requires contention management with simple "error out".